src/test/java/io/aiven/kafka/connect/common/grouper/ParquetTopicPartitionRecordGrouperTest.java

```diff
@@ -0,0 +1,173 @@
+    static final SinkRecord KRT1P1R1 = new SinkRecord(
+            "topic1", 0,
+            SchemaBuilder.string().optional().version(2).build(), null,
+            SchemaBuilder.string().optional().version(2).build(), null, 1001);
+    static final SinkRecord KRT1P1R2 = new SinkRecord(
+            "topic1", 0,
+            SchemaBuilder.string().optional().version(3).build(), null,
+            SchemaBuilder.string().optional().version(3).build(), null, 1002);
+    static final SinkRecord KRT1P1R3 = new SinkRecord(
+            "topic1", 0,
+            SchemaBuilder.string().optional().version(3).build(), "some_key",
+            SchemaBuilder.string().optional().version(3).build(), null, 1003);
+
+    static final TimestampSource DEFAULT_TS_SOURCE =
+            TimestampSource.of(TimestampSource.Type.WALLCLOCK);
+
+    @Test
+    void rotateOnKeySchemaChanged() {
+        final Template filenameTemplate = Template.of("{{topic}}-{{partition}}-{{start_offset}}");
+        final RecordGrouper grouper =
+                new ParquetTopicPartitionRecordGrouper(
+                        filenameTemplate, null, DEFAULT_TS_SOURCE);
+
+        grouper.put(KT0P0R0);
+        grouper.put(KT0P0R1);
+        grouper.put(KT0P0R2);
+        grouper.put(KT0P0R3);
+        grouper.put(KT0P0R4);
+        grouper.put(KT0P0R5);
+
+        final Map<String, List<SinkRecord>> records = grouper.records();
+        assertThat(
+                records.keySet(),
+                containsInAnyOrder("topic0-0-0", "topic0-0-4")
+        );
+        assertThat(records.get("topic0-0-0"),
+                contains(KT0P0R0, KT0P0R1, KT0P0R2, KT0P0R3));
+        assertThat(records.get("topic0-0-4"),
+                contains(KT0P0R4, KT0P0R5));
+    }
+
+    @Test
+    void rotateOnValueSchemaChanged() {
+        final Template filenameTemplate = Template.of("{{topic}}-{{partition}}-{{start_offset}}");
+        final RecordGrouper grouper =
+                new ParquetTopicPartitionRecordGrouper(
+                        filenameTemplate, null, DEFAULT_TS_SOURCE);
+
+        grouper.put(RT0P1R0);
+        grouper.put(RT0P1R1);
+        grouper.put(RT0P1R2);
+        grouper.put(RT0P1R3);
+
+        final Map<String, List<SinkRecord>> records = grouper.records();
+        assertThat(
+                records.keySet(),
+                containsInAnyOrder("topic0-1-10", "topic0-1-12")
+        );
+        assertThat(records.get("topic0-1-10"),
+                contains(RT0P1R0, RT0P1R1));
+        assertThat(records.get("topic0-1-12"),
+                contains(RT0P1R2, RT0P1R3));
+    }
+
+    @Test
+    void rotateOnValueSchemaChangedAndButchSize() {
+        final Template filenameTemplate = Template.of("{{topic}}-{{partition}}-{{start_offset}}");
+        final RecordGrouper grouper =
+                new ParquetTopicPartitionRecordGrouper(
+                        filenameTemplate, 1, DEFAULT_TS_SOURCE);
```

Comment:
Batch size 1 makes it impossible to differentiate if the rotation happens on the schema change, it seems.